README.txt
väri
====

Color representations and representation conversions in Python. _Väri_ means
'color' in Finnish.

install
-------

`pip install vari`. Done.

use
---

Create a `Vari` instance with one of many color representations:

    Vari(rgb=(255, 0, 0))
    Vari('#ff0000')
    Vari('red')

Ask the instance for other representations:

    c = Vari('#f00')
    c.rgb  # (255, 0, 0)
    c.hex  # '#ff0000'
    c.x256  # 196

future
------

- hsl repr


CHANGES.txt
v0.0.1, 2013-02-22 -- RGB and HEX representations.
v0.0.2, 2013-02-22 -- Add ansi x256 8-bit color repr.


tests/context.py
# -*- coding: utf-8 -*-

import sys
import os
sys.path.insert(0, os.path.abspath('..'))

import vari


tests/test_basic.py
# -*- coding: utf-8 -*-

from .context import vari

import unittest


class VariTestSuite(unittest.TestCase):

    def setUp(self):
        self.red = vari.Vari(rgb=(255, 0, 0))

    def test_vari_repr(self):
        self.assertEqual(repr(self.red), '<Vari #ff0000>')

    def test_vari_instance_comparison(self):
        red = vari.Vari(rgb=(255, 0, 0))
        self.assertEqual(self.red, red)

    def test_load_hex(self):
        red = vari.Vari('#ff0000')
        self.assertEqual(self.red, red)

    def test_load_hex_short(self):
        red = vari.Vari('#f00')
        self.assertEqual(self.red, red)

    def test_load_name(self):
        red = vari.Vari('red')
        self.assertEqual(self.red, red)

    def test_2rgb(self):
        self.assertEqual((255, 0, 0), self.red.rgb)

    def test_2hex(self):
        self.assertEqual('#ff0000', self.red.hex)

    def test_2web(self):
        self.assertEqual('red', self.red.web)

    def test_2x256(self):
        self.assertEqual(196, self.red.x256)


if __name__ == '__main__':
    unittest.main()


vari/core.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Phil Adams http://philadams.net

Color representation conversions in Python.
"""

import re

from helpers import COLORS_NAMES

COLORS_RGB = dict((v, k) for k, v in COLORS_NAMES.iteritems())
HEX_COLOR = re.compile(r'^#[0-9a-fA-F]{6}')
HEX_COLOR_SHORT = re.compile(r'^#[0-9a-fA-F]{3}')


def rgb2hex(clr):
    """Convert rgb to hex values.
    - (r,g,b) where each value is 0..255
    - RRGGBB where each value is 00..FF"""
    return '#' + ''.join(map(chr, clr)).encode('hex')


def hex2rgb(clr):
    """Convert hex to rgb values.
    - (r,g,b) where each value is 0..255
    - RRGGBB where each value is 00..FF"""
    return tuple(map(ord, clr[1:].decode('hex')))


def rgb2x256(clr):
    """convert rgb to x256 8-bit color
    - (r,g,b) where each value is 0..255
    - 256 colors in 8 bits RRRGGGBB, returned as single int 0..255"""
    r, g, b = clr
    grey = False
    maybe_gray = True
    for i in range(3, 256, 42):
        if r <= i or g <= i or b <= i:
            grey = r <= i and g <= i and b <= i
            break
    if grey:
        color = 232 + int(sum(clr) / 33.0)
    else:
        components = ((r, 36), (g, 6), (b, 1))
        color = sum([16] + [int(6*val/256.0)*bmp for val, bmp in components])
    return color


class Vari(object):
    """Color representation and representation conversion.
    In Finnish, väri means color."""

    def __init__(self, color=None, **kwargs):

        self._rgb = None  # internal repr
        self.update(color, **kwargs)

        # if we've not set self._rgb, we couldn't init the color
        if not self._rgb:
            raise ValueError('cannot initialize Vari instance')

    def update(self, color, **kwargs):

        # handle kwargs[rgb]
        if 'rgb' in kwargs:
            self._rgb = tuple(map(int, kwargs['rgb']))

        elif color:

            # handle hex
            if color.startswith('#'):
                if HEX_COLOR.match(color):
                    self._rgb = hex2rgb(color)
                elif HEX_COLOR_SHORT.match(color):
                    clr = '#' + ''.join(['%s%s' % (e,e) for e in color[1:]])
                    self._rgb = hex2rgb(clr)

            # handle web
            elif color in COLORS_NAMES:
                self._rgb = COLORS_NAMES[color]


    def __eq__(self, other):
        return self._rgb == other._rgb

    def __repr__(self):
        return '<Vari %s>' % self.hex

    @property
    def rgb(self):
        return self._rgb

    @rgb.setter
    def rgb(self, rgb):
        self._rgb = rgb

    @property
    def hex(self):
        return rgb2hex(self._rgb)

    @property
    def web(self):
        return COLORS_RGB.get(self._rgb, self.hex)

    @property
    def x256(self):
        return rgb2x256(self._rgb)

def cli():
    import argparse

    # populate and parse command line options
    desc = 'Color representation conversions in Python.'
    desc += '\nhttp://github.com/philadams/vari'
    parser = argparse.ArgumentParser(description=desc)
    args = parser.parse_args()

    print('commandline tool coming soon...')


vari/helpers.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Phil Adams http://philadams.net

Color representation conversions in Python - helpers.
"""

COLORS_NAMES = {'snow': (255, 250, 250),
                'ghostwhite': (248, 248, 255),
                'whitesmoke': (245, 245, 245),
                'gainsboro': (220, 220, 220),
                'floralwhite': (255, 250, 240),
                'oldlace': (253, 245, 230),
                'linen': (250, 240, 230),
                'antiquewhite': (250, 235, 215),
                'papayawhip': (255, 239, 213),
                'blanchedalmond': (255, 235, 205),
                'bisque': (255, 228, 196),
                'peachpuff': (255, 218, 185),
                'navajowhite': (255, 222, 173),
                'moccasin': (255, 228, 181),
                'cornsilk': (255, 248, 220),
                'ivory': (255, 255, 240),
                'lemonchiffon': (255, 250, 205),
                'seashell': (255, 245, 238),
                'honeydew': (240, 255, 240),
                'mintcream': (245, 255, 250),
                'azure': (240, 255, 255),
                'aliceblue': (240, 248, 255),
                'lavender': (230, 230, 250),
                'lavenderblush': (255, 240, 245),
                'mistyrose': (255, 228, 225),
                'white': (255, 255, 255),
                'black': (0, 0, 0),
                'darkslategray': (47, 79, 79),
                'darkslategrey': (47, 79, 79),
                'dimgray': (105, 105, 105),
                'dimgrey': (105, 105, 105),
                'slategray': (112, 128, 144),
                'slategrey': (112, 128, 144),
                'lightslategray': (119, 136, 153),
                'lightslategrey': (119, 136, 153),
                'gray': (190, 190, 190),
                'grey': (190, 190, 190),
                'lightgrey': (211, 211, 211),
                'lightgray': (211, 211, 211),
                'midnightblue': (25, 25, 112),
                'navy': (0, 0, 128),
                'navyblue': (0, 0, 128),
                'cornflowerblue': (100, 149, 237),
                'darkslateblue': (72, 61, 139),
                'slateblue': (106, 90, 205),
                'mediumslateblue': (123, 104, 238),
                'lightslateblue': (132, 112, 255),
                'mediumblue': (0, 0, 205),
                'royalblue': (65, 105, 225),
                'blue': (0, 0, 255),
                'dodgerblue': (30, 144, 255),
                'deepskyblue': (0, 191, 255),
                'skyblue': (135, 206, 235),
                'lightskyblue': (135, 206, 250),
                'steelblue': (70, 130, 180),
                'lightsteelblue': (176, 196, 222),
                'lightblue': (173, 216, 230),
                'powderblue': (176, 224, 230),
                'paleturquoise': (175, 238, 238),
                'darkturquoise': (0, 206, 209),
                'mediumturquoise': (72, 209, 204),
                'turquoise': (64, 224, 208),
                'cyan': (0, 255, 255),
                'lightcyan': (224, 255, 255),
                'cadetblue': (95, 158, 160),
                'mediumaquamarine': (102, 205, 170),
                'aquamarine': (127, 255, 212),
                'darkgreen': (0, 100, 0),
                'darkolivegreen': (85, 107, 47),
                'darkseagreen': (143, 188, 143),
                'seagreen': (46, 139, 87),
                'mediumseagreen': (60, 179, 113),
                'lightseagreen': (32, 178, 170),
                'palegreen': (152, 251, 152),
                'springgreen': (0, 255, 127),
                'lawngreen': (124, 252, 0),
                'green': (0, 255, 0),
                'chartreuse': (127, 255, 0),
                'mediumspringgreen': (0, 250, 154),
                'greenyellow': (173, 255, 47),
                'limegreen': (50, 205, 50),
                'yellowgreen': (154, 205, 50),
                'forestgreen': (34, 139, 34),
                'olivedrab': (107, 142, 35),
                'darkkhaki': (189, 183, 107),
                'khaki': (240, 230, 140),
                'palegoldenrod': (238, 232, 170),
                'lightgoldenrodyellow': (250, 250, 210),
                'lightyellow': (255, 255, 224),
                'yellow': (255, 255, 0),
                'gold': (255, 215, 0),
                'lightgoldenrod': (238, 221, 130),
                'goldenrod': (218, 165, 32),
                'darkgoldenrod': (184, 134, 11),
                'rosybrown': (188, 143, 143),
                'indianred': (205, 92, 92),
                'saddlebrown': (139, 69, 19),
                'sienna': (160, 82, 45),
                'peru': (205, 133, 63),
                'burlywood': (222, 184, 135),
                'beige': (245, 245, 220),
                'wheat': (245, 222, 179),
                'sandybrown': (244, 164, 96),
                'tan': (210, 180, 140),
                'chocolate': (210, 105, 30),
                'firebrick': (178, 34, 34),
                'brown': (165, 42, 42),
                'darksalmon': (233, 150, 122),
                'salmon': (250, 128, 114),
                'lightsalmon': (255, 160, 122),
                'orange': (255, 165, 0),
                'darkorange': (255, 140, 0),
                'coral': (255, 127, 80),
                'lightcoral': (240, 128, 128),
                'tomato': (255, 99, 71),
                'orangered': (255, 69, 0),
                'red': (255, 0, 0),
                'hotpink': (255, 105, 180),
                'deeppink': (255, 20, 147),
                'pink': (255, 192, 203),
                'lightpink': (255, 182, 193),
                'palevioletred': (219, 112, 147),
                'maroon': (176, 48, 96),
                'mediumvioletred': (199, 21, 133),
                'violetred': (208, 32, 144),
                'magenta': (255, 0, 255),
                'violet': (238, 130, 238),
                'plum': (221, 160, 221),
                'orchid': (218, 112, 214),
                'mediumorchid': (186, 85, 211),
                'darkorchid': (153, 50, 204),
                'darkviolet': (148, 0, 211),
                'blueviolet': (138, 43, 226),
                'purple': (160, 32, 240),
                'mediumpurple': (147, 112, 219),
                'thistle': (216, 191, 216),
                'darkgrey': (169, 169, 169),
                'darkgray': (169, 169, 169),
                'darkblue': (0, 0, 139),
                'darkcyan': (0, 139, 139),
                'darkmagenta': (139, 0, 139),
                'darkred': (139, 0, 0),
                'lightgreen': (144, 238, 144)}
